src/pcre2_jit_compile.c

```diff
@@ -5888,29 +5990,32 @@ while (TRUE)
     case OP_NOTEXACT:
     case OP_NOTEXACTI:
 #if defined SUPPORT_UNICODE && PCRE2_CODE_UNIT_WIDTH != 32
-    if (common->utf) return consumed;
+    if (common->utf)
+      {
+      chars_end = chars;
+      continue;
+      }
 #endif
     any = TRUE;
     repeat = GET2(cc, 1);
     cc += 1 + IMM2_SIZE + 1;
     break;
 
     default:
-    return consumed;
+    chars_end = chars;
+    continue;
     }
 
+  SLJIT_ASSERT(chars < chars_end);
+
   if (any)
     {
     do
       {
       chars->count = 255;
-
-      consumed++;
-      if (--max_chars == 0)
-        return consumed;
       chars++;
       }
-    while (--repeat > 0);
+    while (--repeat > 0 && chars < chars_end);
 
     repeat = 1;
```

Comment:
There are several places that do `repeat = 1`, and I think you need to remember to do that in all the places that do a `continue`. That seems error-prone. Couldn't we move all the `repeat = 1` assignments up to the start of the loop, where last/caseless/etc are initialized?

Reply:
Unfortunately not.

https://github.com/PCRE2Project/pcre2/blob/master/src/pcre2_jit_compile.c#L5883

For combined opcodes, we just record the repeat, and continue the main loop. Then we process the actual opcode with the recorded repeat.

E.g.: OP_TYPEEXACT followed by OP_ANY.

Reply:
Maybe a goto could be used here, but I try to not use goto when possible.